Removing Recheck and Filter Condition's columns from Bitmap Heap Scan output list when using ORCA.

InnerLife0 opened the pull request:
The changes introduced in PR #12176 revealed a problem with excessive Bitmap Heap Scan's targetlist which includes Recheck and Filter Condition's columns (introduced in old 97edf9b).
Memory tuple binding generated based on targetlist with different count of attrs may point to different offsets and so, to some garbage or out of range memory area. See issue #12796 for more.
Additional regression test shows how to reproduce a bug before fix. A couple of tests was modified, because there is no need of projection in them started from this patch.

Stolb27 commented:
@maksm90, please take a look at this one.

Stolb27 commented on src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -1262,8 +1255,7 @@ CTranslatorExprToDXL::PdxlnDynamicBitmapTableScan(
 	CColRefSet *pcrsOutput = pexprScan->Prpp()->PcrsRequired();
-	AddBitmapFilterColumns(m_mp, pop, pexprRecheckCond, pexprScalar,
-						   pcrsOutput);
+	AddBitmapFilterColumns(m_mp, pop, pexprScalar, pcrsOutput);
```
I've tried to validate an original commit https://github.com/arenadata/gpdb/commit/97edf9b1cbd47c33918e5780e04ea241b59a2f46 that implements this behavior. I found that this patch doesn't affect any `DynamicBitmapHeapScan` tests. So I tried to find out why. There is an appropriate [test case](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/gporca/data/dxl/minidump/sql/DynamicBitmapIndexScan.sql) in the original commit. This test case helped me after small corrections (including changing value of `allow_system_table_mods` and shrinking namespace filters) to generate a plan with an excessive column `flex_value_id` in the target list:

```
test=# explain (costs off, verbose) select inner_table.flex_value_set_id from inner_table where inner_table.flex_value_id = 1;
                                      QUERY PLAN                                       
---------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   Output: flex_value_set_id
   ->  Result
         Output: flex_value_set_id
         ->  Sequence
               Output: flex_value_set_id, flex_value_id
               ->  Partition Selector for inner_table (dynamic scan id: 1)
                     Partitions selected: 6 (out of 6)
               ->  Dynamic Bitmap Heap Scan on public.inner_table (dynamic scan id: 1)
                     Output: flex_value_set_id, flex_value_id
                     Recheck Cond: (inner_table.flex_value_id = 1::numeric)
                     ->  Dynamic Bitmap Index Scan on inner_table_idx
                           Index Cond: (inner_table.flex_value_id = 1::numeric)
 Optimizer: Pivotal Optimizer (GPORCA)
```

But when I tried to combine Dynamic Scan with Shared Scan, I got a valid plan with projection node before materialization:

```
test=# explain (costs off, verbose) with mat as(select inner_table.flex_value_set_id i from inner_table where inner_table.flex_value_id = 1) select m1.i from mat m1 join mat m2 on m1.i = m2.i;
                                               QUERY PLAN                                                
---------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)
   Output: share0_ref3.flex_value_set_id
   ->  Sequence
         Output: share0_ref3.flex_value_set_id
         ->  Shared Scan (share slice:id 2:0)
               Output: share0_ref1.flex_value_set_id
               ->  Materialize
                     Output: inner_table.flex_value_set_id
                     ->  Result
                           Output: inner_table.flex_value_set_id
                           ->  Sequence
                                 Output: inner_table.flex_value_set_id, inner_table.flex_value_id
                                 ->  Partition Selector for inner_table (dynamic scan id: 1)
                                       Partitions selected: 6 (out of 6)
                                 ->  Dynamic Bitmap Heap Scan on public.inner_table (dynamic scan id: 1)
                                       Output: inner_table.flex_value_set_id, inner_table.flex_value_id
                                       Recheck Cond: (inner_table.flex_value_id = 1::numeric)
                                       ->  Dynamic Bitmap Index Scan on inner_table_idx
                                             Index Cond: (inner_table.flex_value_id = 1::numeric)
         ->  Nested Loop
               Output: share0_ref3.flex_value_set_id
...
```

maksm90 commented:
Could anyone assume why the `AddBitmapFilterColumns()` call is needed in general if we get `pcrsOutput` list of attributes from upper nodes or compute it via `PcrsRequired()` method of BitmapTableScan node?

maksm90 commented on src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -967,12 +967,6 @@ CTranslatorExprToDXL::AddBitmapFilterColumns(
-		// add the columns used in the recheck condition
-		pcrsAdditional->Include(pexprRecheckCond->DeriveUsedColumns());
-	}
-
```
Excluding attributes that are from recheck condition and at the same time retaining those ones from filter is not solution, in general. We might find an example exposing redundant attributes to the output list of BitmapHeapScan that are member of filter condition but not ones in target list, e.g. the following:
```sql
create table material_bitmapscan_1(i int, j int, k timestamp, l timestamp) with(appendonly=true) distributed replicated;
create index material_bitmapscan_1_idx on material_bitmapscan_1 using btree(k);
insert into material_bitmapscan_1
select i, mod(i, 10), timestamp '2021-06-01' + interval '1' day * mod(i, 30), timestamp '2021-06-01' + interval '1' day * mod(i, 30)
from generate_series(1, 10000) i;

explain (costs off, verbose) with mat as(
    select i, j from material_bitmapscan_1
    where i = 2 and j = 2 and k = timestamp '2021-06-03' and l = timestamp '2021-06-03'
)
select m1.i
from mat m1 join mat m2 on m1.j = m2.j;
                                                                                       QUERY PLAN                                                                                       
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)
   Output: share0_ref3.i
   ->  Sequence
         Output: share0_ref3.i
         ->  Shared Scan (share slice:id 1:0)
               Output: share0_ref1.i, share0_ref1.j
               ->  Materialize
                     Output: material_bitmapscan_1.i, material_bitmapscan_1.j, material_bitmapscan_1.l
                     ->  Bitmap Heap Scan on public.material_bitmapscan_1
                           Output: material_bitmapscan_1.i, material_bitmapscan_1.j, material_bitmapscan_1.l
                           Recheck Cond: (material_bitmapscan_1.k = '2021-06-03 00:00:00'::timestamp without time zone)
                           Filter: ((material_bitmapscan_1.i = 2) AND (material_bitmapscan_1.j = 2) AND (material_bitmapscan_1.l = '2021-06-03 00:00:00'::timestamp without time zone))
                           ->  Bitmap Index Scan on material_bitmapscan_1_idx
                                 Index Cond: (material_bitmapscan_1.k = '2021-06-03 00:00:00'::timestamp without time zone)
         ->  Hash Join
               Output: share0_ref3.i
               Hash Cond: (share0_ref3.j = share0_ref2.j)
               ->  Result
                     Output: share0_ref3.i, share0_ref3.j
                     Filter: (share0_ref3.j = 2)
                     ->  Shared Scan (share slice:id 1:0)
                           Output: share0_ref3.i, share0_ref3.j
               ->  Hash
                     Output: share0_ref2.j
                     ->  Result
                           Output: share0_ref2.j
                           Filter: (share0_ref2.j = 2)
                           ->  Shared Scan (share slice:id 1:0)
                                 Output: share0_ref2.i, share0_ref2.j
 Optimizer: Pivotal Optimizer (GPORCA)

with mat as(
    select i, j from material_bitmapscan_1
    where i = 2 and j = 2 and k = timestamp '2021-06-03' and l = timestamp '2021-06-03'
)
select m1.i
from mat m1 join mat m2 on m1.j = m2.j;
 i 
---
(0 rows)
```

I might propose to remove a whole `AddBitmapFilterColumns()` call from `PdxlnBitmapTableScan()` one but not sure about negative side effects there are.

Commit:
Removing Recheck and Filter Condition's columns from Bitmap Heap Scan output list when using ORCA.
The changes introduced in PR #12176 revealed a problem with excessive Bitmap Heap Scan's targetlist which includes Recheck and Filter Condition's columns (introduced in old 97edf9b).
Memory tuple binding generated based on targetlist with different count of attrs may point to different offsets and so, to some garbage or out of range memory area. See issue #12796 for more.
Additional regression test shows how to reproduce a bug before fix. A couple of tests was modified, because there is no need of projection in them started from this patch.

Commit:
Natively raise exception in case of targetlist mismatch.

maksm90 commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -3608,6 +3608,11 @@ CTranslatorDXLToPlStmt::TranslateDXLCTEProducerToSharedScan(
 	}
 
+	// targetlist mismatch leads to different tuple bindings, see #12796
+	if (list_length(child_plan->targetlist) != list_length(plan->targetlist))
```
Is it correct to compare just size (not content, order) of target lists?

InnerLife0 replied:
I don't see any valid situation when parent node may has, for example `i, j, k` columns, when child node has `i, j, l` columns or wrong sorted columns. If there are some situations when one set does not include the other it should be hooked and fixed somewhere outside `TranslateDXLCTEProducerToSharedScan` function.

maksm90 replied:
OK then point in comment that in current plan node we expect target list as sub-list from child one because it's not evident that comparing sizes we compare lists in a whole.

maksm90 commented on src/backend/gporca/data/dxl/minidump/Gist-AOCOTable-NonLossy-BitmapIndexPlan.mdp:
```diff
@@ -282,65 +282,47 @@ Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..0.00 rows=1 width=4)
+        </dxl:Sort>
       </dxl:GatherMotion>
     </dxl:Plan>
   </dxl:Thread>
```
please, fix query plan in plain text format in the comment section: remove `Result` node and refine cost values

InnerLife0 replied:
Thanks

InnerLife0 replied:
I should say other tests has different plan too, but their plan in comment section was not touched, because the difference was not caused by current patch.
As told on the one of our voice sessions, the problem of not actual mdp files is global and also related to their bodies.

maksm90 commented on src/test/regress/expected/gporca_optimizer.out:
```diff
@@ -13729,7 +13729,7 @@ and first_id in (select first_id from mat_w);
  Optimizer: Pivotal Optimizer (GPORCA)
- Settings: enable_seqscan=on, optimizer=on
+ Settings: optimizer=on
 (31 rows)
```
Why not change the above test case to verify current patch, e.g. replace the type of `second_id` from integer to some more complex (timestamp in your provided case)?

InnerLife0 replied:
This test was introduced to show a different issue. Yes, this issue is related to ours, but it's different, so I don't think it's correct to modify existed test.
The test related to our issue was introduced by this patch separately.

Commit:
Commenting. mdp comment plan update.

Stolb27 committed:
Merge branch 'adb-6.x' into ADBDEV-2204-3
